craft_parts/executor/part_handler.py

```diff
@@ -655,6 +733,35 @@ def _remove(filename: Path) -> None:
         shutil.rmtree(filename)
 
 
+def _apply_file_filter(
+    *, filter_files: Set[str], filter_dirs: Set[str], destdir: Path
+) -> None:
+    for (root, directories, files) in os.walk(destdir, topdown=True):
+        for file_name in files:
+            path = Path(root, file_name)
+            relpath = path.relative_to(destdir)
+            if str(relpath) not in filter_files and not overlays.is_whiteout_file(path):
+                logger.debug("delete file: %s", relpath)
+                path.unlink()
+
+        for directory in directories:
+            path = Path(root, directory)
+            relpath = path.relative_to(destdir)
+            if path.is_symlink():
+                if str(relpath) not in filter_files:
+                    logger.debug("delete symlink: %s", relpath)
+                    path.unlink()
+            elif str(relpath) not in filter_dirs:
+                logger.debug("delete dir: %s", relpath)
+                # Don't descend into this directory-- we'll just delete it
+                # entirely.
+                directories.remove(directory)
+                if path.is_dir():
```

Comment:
As we're iterating `directories` above, how this could not be a dir?

Reply:
Good catch, this seems to be a leftover from the original code before we tested for symbolic links.

Reply:
Fixed to remove redundant testing.